goi18n/constants_command.go

```diff
@@ -0,0 +1,237 @@
+}
+
+func (cc *constantsCommand) parse(arguments []string) {
+	constantsCmd := flag.NewFlagSet("constants", flag.ExitOnError)
+	constantsCmd.Usage = usageConstants
+
+	packageName := constantsCmd.String("package", "R", "")
+	outdir := constantsCmd.String("outdir", ".", "")
+
+	constantsCmd.Parse(arguments)
+
+	if constantsCmd.Parsed() {
+		cc.translationFiles = constantsCmd.Args()
+		cc.packageName = *packageName
+		cc.outdir = *outdir
+	} else {
+		usageConstants()
+	}
+}
+
+func (cc *constantsCommand) SetArgs(args []string) {
+	cc.translationFiles = args
+}
+
+func usageConstants() {
+	fmt.Printf(`Generate constant file from translation file.
+
+Usage:
+
+    goi18n constants [options] [file]
+
+Translation files:
+
+    A translation file contains the strings and translations for a single language.
+
+    Translation file names must have a suffix of a supported format (e.g. .json) and
+    contain a valid language tag as defined by RFC 5646 (e.g. en-us, fr, zh-hant, etc.).
+
+Options:
+
+    -package name
+        goi18n generates the constant file under the package name.
+        Default: R
+
+    -outdir directory
+        goi18n writes the constant file to this directory.
+        Default: .
+
+`)
+	os.Exit(1)
+}
+
+// commonInitialisms is a set of common initialisms.
+// Only add entries that are highly unlikely to be non-initialisms.
+// For instance, "ID" is fine (Freudian code is rare), but "AND" is not.
+// https://github.com/golang/lint/blob/master/lint.go
+var commonInitialisms = map[string]bool{
+	"API":   true,
+	"ASCII": true,
+	"CPU":   true,
+	"CSS":   true,
+	"DNS":   true,
+	"EOF":   true,
+	"GUID":  true,
+	"HTML":  true,
+	"HTTP":  true,
+	"HTTPS": true,
+	"ID":    true,
+	"IP":    true,
+	"JSON":  true,
+	"LHS":   true,
+	"QPS":   true,
+	"RAM":   true,
+	"RHS":   true,
+	"RPC":   true,
+	"SLA":   true,
+	"SMTP":  true,
+	"SQL":   true,
+	"SSH":   true,
+	"TCP":   true,
+	"TLS":   true,
+	"TTL":   true,
+	"UDP":   true,
+	"UI":    true,
+	"UID":   true,
+	"UUID":  true,
+	"URI":   true,
+	"URL":   true,
+	"UTF8":  true,
+	"VM":    true,
+	"XML":   true,
+	"XSRF":  true,
+	"XSS":   true,
+}
+
+func toCamelCase(id string) string {
+	var result string
+
+	r := regexp.MustCompile(`[\-\.\_\s]`)
+	words := r.Split(id, -1)
+
+	for _, w := range words {
+		upper := strings.ToUpper(w)
+		if commonInitialisms[upper] {
+			result += upper
+			continue
+		}
+
+		if len(w) > 0 {
+			u := []rune(w)
+			u[0] = unicode.ToUpper(u[0])
+			result += string(u)
+		}
+	}
+	return result
+}
+
+func toComment(trans translation.Translation) string {
+	var result string
+	data := trans.MarshalInterface().(map[string]interface{})
+
+	t := data["translation"]
+
+	switch v := reflect.ValueOf(t); v.Kind() {
+	case reflect.Map:
+		for _, k := range []language.Plural{"zero", "one", "two", "few", "many", "other"} {
+			vt := v.MapIndex(reflect.ValueOf(k))
+			if !vt.IsValid() {
+				continue
+			}
+
+			if len(result) > 0 {
+				result += "\n// "
```

Comment:
It would be better if this sort of formatting were in the template itself.

For example

```
type templateConstants struct {
    ID      string
    Name    string
    Comments []string
}
```

Then in the template you can range over `Comments`.